models/src/main/scala/coop/rchain/models/rholangn/parmanager/Serialization.scala

```diff
@@ -2,97 +2,69 @@ package coop.rchain.models.rholangn.parmanager
     writePar(kv._1) *> writePar(kv._2)
 
-  def writeTupleStringPar(kv: (String, RhoTypeN)): Eval[Unit] =
+  def writeTupleStringPar(
+      kv: (String, RhoTypeN)
+  )(implicit rec: RhoTypeN => Eval[Unit]): Eval[Unit] =
     writer.write(kv._1) *> writePar(kv._2)
 
   // Writes serialized value of a sequence
   def writeSeq[T](seq: Seq[T], f: T => Eval[Unit]): Eval[Unit] =
     writer.write(seq.size) *> seq.traverse_(f)
 }
 
-/** Wrapper for protobuf de-serialization of primitive types. */
-private class ProtobufReader(input: CodedInputStream) {
-  // NOTE: Eval.always is used to ensure correct deserialization and read from input stream
-  def readByte: Eval[Byte]         = Eval.always(input.readRawByte())
-  def readBytes: Eval[Array[Byte]] = Eval.always(input.readByteArray())
-  def readBool: Eval[Boolean]      = Eval.always(input.readBool())
-  def readInt: Eval[Int]           = Eval.always(input.readUInt32())
-  def readLong: Eval[Long]         = Eval.always(input.readUInt64())
-  def readString: Eval[String]     = Eval.always(input.readString())
-}
-
 object Serialization {
 
-  // TODO: Properly handle errors
-  @SuppressWarnings(Array("org.wartremover.warts.TryPartial"))
-  def serializeToBytes(par: RhoTypeN): Eval[Array[Byte]] =
-    par.serializedSize.flatMap { serSize =>
-      Using(new ByteArrayOutputStream(serSize)) { baos =>
-        Serialization.serialize(par, baos).map { _ =>
-          baos.flush()
-          baos.toByteArray
-        }
-      }.get
-    }
-
-  // TODO: Properly handle errors
-  @SuppressWarnings(Array("org.wartremover.warts.TryPartial"))
-  def deserializeFromBytes(bv: Array[Byte]): ParN =
-    Using(new ByteArrayInputStream(bv))(Serialization.deserialize(_).value).get
-
   // TODO: Properly handle errors with return type (remove throw)
-  def serialize(par: RhoTypeN, output: OutputStream): Eval[Unit] = {
-    val cos         = CodedOutputStream.newInstance(output)
-    val protoWriter = new ProtobufPrimitiveWriter(cos)
-
-    // Serializer with recursive traversal of the whole object at once
-    lazy val serializer = new ProtobufRecWriter(protoWriter, writeRec)
-
-    // Serializer with recursive traversal using memoized values
-    // val serializer = new ProtobufRecWriter(protoWriter, _.serialized.flatMap(protoWriter.writeRaw))
+  def serialize(par: RhoTypeN, primitiveWriter: PrimitiveWriter): Eval[Unit] = {
 
     @SuppressWarnings(Array("org.wartremover.warts.Throw"))
     def writeRec(p: RhoTypeN): Eval[Unit] = Eval.defer {
-      import protoWriter._
-      import serializer._
+      import primitiveWriter._
+      val recWriterOps = new RhoTypeWriterOps(primitiveWriter)
+      import recWriterOps._
+
+      // Recursive traversal of the whole object at once
+      implicit val rec: RhoTypeN => Eval[Unit] = writeRec
+      // Recursive traversal using memoized values
+      // implicit val rec: RhoTypeN => Eval[Unit] = _.serialized.flatMap(primitiveWriter.writeRaw)
```

Comment:
Why all this complication with implicit function?

The point of different ways to serialize is to be able to select it on top call. `serialize` function should have control how serialization is done. We want to be able to call both in the same source code.

Reply:
Might be using implicit is not the best way here, I'll make an explicit dependency.
What I'm trying to so is to separate ParN traversal from writing to wire. So traversal is separate and serialisation is an implementation for processing of primitives while traversing.

Reply:
It's already separated by PrimitiveWriter and PrimitiveReader.